File: solutions/typescript/src/TaskManagement/TaskManagementSystem.ts
import Task from "./Task";
import { TaskPriorityEnum } from "./TaskPriorityEnum";
import { TaskStatusEnum } from "./TaskStatusEnum";
import User from "./User";
import TaskList from "./TaskList";
import Comment from "./Comment";
import TaskSortingStrategy from "./SortingStrategy/TaskSortingStrategy";

export default class TaskManagementSystem {
  private static instance: TaskManagementSystem | null;
  private users: Map<String, User>;
  private tasks: Map<String, Task>;
  private taskLists: Map<String, TaskList>;

  private constructor() {
    this.users = new Map();
    this.tasks = new Map();
    this.taskLists = new Map();
  }

  static getInstance() {
    if (!TaskManagementSystem.instance) {
      TaskManagementSystem.instance = new TaskManagementSystem();
    }
    return TaskManagementSystem.instance;
  }

  createUser(name: string, email: string) {
    const user = new User(name, email);
    this.users.set(user.getId(), user);
    return user;
  }

  createTaskList(listName: string) {
    const taskList = new TaskList(listName);
    this.taskLists.set(taskList.getId(), taskList);
    return taskList;
  }

  createTask(
    listId: string,
    title: string,
    description: string,
    dueDate: Date,
    priority: TaskPriorityEnum,
    createdByUserId: string,
  ) {
    const taskList = this.taskLists.get(listId);
    if (taskList == null) throw new Error("TaskList not found.");

    const createdBy = this.users.get(createdByUserId);
    if (createdBy == null) throw new Error("User not found.");

    const task = new Task(title, description, dueDate, priority, createdBy);

    this.tasks.set(task.getId(), task);
    taskList.addTask(task);
    return task;
  }

  getTaskById(taskId: string) {
    const task = this.tasks.get(taskId);
    if (!task) {
      throw new Error("Task not found: " + taskId);
    }
    return task;
  }

  updateTaskStatus(taskId: string, status: TaskStatusEnum) {
    this.getTaskById(taskId).updateStatus(status);
  }

  updateTaskPriority(taskId: string, priority: TaskPriorityEnum) {
    this.getTaskById(taskId).updatePriority(priority);
  }

  assignTask(taskId: string, userId: string) {
    const user = this.users.get(userId);
    if (user == null) throw new Error("User not found.");
    this.getTaskById(taskId).assignUser(user);
  }

  addComment(taskId: string, commentText: string, author: User) {
    const task = this.getTaskById(taskId);
    task.addComment(new Comment(commentText, author));
  }

  listTasksByUser(userId: string) {
    return [...this.tasks.values()].filter(
      (task) => userId == task.getAssignedTo()?.getId(),
    );
  }

  listTasksByStatus(status: TaskStatusEnum) {
    return [...this.tasks.values()].filter(
      (task) => task.getStatus() == status,
    );
  }

  deleteTask(taskId: string) {
    this.tasks.delete(taskId);
  }

  searchTasks(keyword: string, sortingStrategy: TaskSortingStrategy) {
    const matchingTasks = [];
    for (const task of this.tasks.values()) {
      if (
        task.getTitle().includes(keyword) ||
        task.getDescription().includes(keyword)
      ) {
        matchingTasks.push(task);
      }
    }
    sortingStrategy.sort(matchingTasks);
    return matchingTasks;
  }
}

File: solutions/python/taskmanagementsystem/task_management_system.py
import threading
from typing import Dict, List
from user import User
from task import Task
from task_list import TaskList
from task_priority import TaskPriority
from task_status import TaskStatus
from task_sort_strategy import TaskSortStrategy
from task_observer import ActivityLogger
from datetime import date

class TaskManagementSystem:
    _instance = None
    _lock = threading.Lock()
    
    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if not self._initialized:
            self._users: Dict[str, User] = {}
            self._tasks: Dict[str, Task] = {}
            self._task_lists: Dict[str, TaskList] = {}
            self._initialized = True
    
    @classmethod
    def get_instance(cls):
        return cls()
    
    def create_user(self, name: str, email: str) -> User:
        user = User(name, email)
        self._users[user.id] = user
        return user
    
    def create_task_list(self, list_name: str) -> TaskList:
        task_list = TaskList(list_name)
        self._task_lists[task_list.id] = task_list
        return task_list
    
    def create_task(self, title: str, description: str, due_date: date,
                   priority: TaskPriority, created_by_user_id: str) -> Task:
        created_by = self._users.get(created_by_user_id)
        if created_by is None:
            raise ValueError("User not found.")
        
        task = Task.TaskBuilder(title) \
            .description(description) \
            .due_date(due_date) \
            .priority(priority) \
            .created_by(created_by) \
            .build()
        
        task.add_observer(ActivityLogger())
        
        self._tasks[task.get_id()] = task
        return task
    
    def list_tasks_by_user(self, user_id: str) -> List[Task]:
        user = self._users.get(user_id)
        return [task for task in self._tasks.values() 
                if task.get_assignee() == user]
    
    def list_tasks_by_status(self, status: TaskStatus) -> List[Task]:
        return [task for task in self._tasks.values() 
                if task.get_status() == status]
    
    def delete_task(self, task_id: str):
        if task_id in self._tasks:
            del self._tasks[task_id]
    
    def search_tasks(self, keyword: str, sorting_strategy: TaskSortStrategy) -> List[Task]:
        matching_tasks = []
        for task in self._tasks.values():
            if (keyword in task.get_title() or 
                keyword in task.get_description()):
                matching_tasks.append(task)
        
        sorting_strategy.sort(matching_tasks)
        return matching_tasks